### PR TITLE
Implement PowerTable as ordered map

### PR DESCRIPTION
**Pull request by ranchalp**
EDIT: builds upon #34 , so merge that one first.

Implements `PowerTable` as an ordered map. To be used by explicit justification to create/verify bitsets (see [here](https://github.com/filecoin-project/go-f3/pull/31#discussion_r1446689822)) and, even if using implicit justification, by the creation/verification of PoFs as part of the catch-up mechanism.

### CODE REVIEW

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:59]: You can remove this, because you can check for duplication after doing the `sort.Search`.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:16]: Please be more clear about what this is indexing. "Maps actor ID to the index of the associated entry in Entries"

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:55]:
```suggestion
// Add adds a new entry to the PowerTable. 
// This is linear in the table size, so adding many entries this way is inefficient.
```

Which is ok because this is only for testing.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:76]: Suggest making this `Get(id ActorID) (Power, []byte)` and return both the power and private key, when you add that. And maybe even add the `ok bool` return here too (otherwise add a `Has()` method and panic here if !ok)

**Review by anorth — COMMENTED** [review]
We really need two things here:
1. A data structure that can be serialized as the canonical power table, giving a CID
2. A runtime structure for looking up power values and keys

This is good for (2), but we won't want to serialise the lookup table. For now lets ignore that problem. We can chose later to add a new type for the serializable form, or manually implement serialisation here, including only the wanted data.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:49]: Done by rebasing from #34

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:15]: It's actually in order by (Power descending, ID ascending), but done otherwise.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:9]: Done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:10]: done

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:29]: I changed it to `[]PowerEntry`.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:59]: Done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:16]: Done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:76]: Done.

**Comment by ranchalp** [comment]
Thanks! addressed the comments :)

File: f3/granite.go
Comment (anorth) [inline f3/granite.go:593]: No need to multiply by one

File: f3/granite.go
Comment (anorth) [inline f3/granite.go:631]: Can you factor out a `hasStrongQuorum(part, total *StoragePower) bool` pure function to use here and above? And maybe do a `hasWeakQuorum` too for symmetry.

File: f3/granite.go
Comment (anorth) [inline f3/granite.go:523]: It's tricky to know how to reference Go big.Ints properly. Is there a good reason this one is not a pointer but the others are? Using a pointer seems to be canonical so I'd suggest doing that here too, unless there's a good reason not to.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:25]: Please note the function takes ownership of the slice - it must not be modified afterwards.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:27]: Extract the repeated `entries[i].Power.Cmp(entries[j].Power)` to a temp.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:51]: Actually please factor out this whole comparison function so it can be re-used between methods.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:55]: You only need to check the exact entry pointed to.

> Search uses binary search to find and return the **smallest** index i in [0, n) at which f(i) is true

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:36]: Check the map size here to ensure there were no duplicates in the input.

File: f3/powertable.go
Comment (anorth) [inline f3/powertable.go:70]: Having seen this in use (especially ignoring the bool return), I have a further suggestion.

The existing code used the zero value map lookup to silently accept messages from participants with no power. Now is the time to change that. I recommend:
- Add a `Has(id)` method here, and call it in `isValid` to reject messages from participants absent or with zero power
- Remove the bool return here and just panic if not found. All the call sites should be able to rely on the prior filtering of valid messages.

File: f3/types.go
Comment (anorth) [inline f3/types.go:10]: Make this a type alias and then you can remove all the wrapper methods.

`type StoragePower  = big.Int`

This is slightly less "safe" in that it doesn't distinguish between different things a big.Int might be representing, but I think it's fine here (and Lotus does too).

File: sim/sim.go
Comment (anorth) [inline sim/sim.go:41]: Let's put a non-empty key in here so that the fake signature verification at least checks it's the right signature. Serialize the participant ID into a byte array, and maybe add some magic bytes to distinguish it from other serialized integers.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:593]: Oops, AI help was too obvious here 😅

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:631]: done.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:523]: You are absolutely right. I started with no pointers but missed this one when switching to pointers.

Comment (ranchalp) [reply f3/granite.go:523]: Done

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:25]: Correct, but not yet an issue. I reckon we leave as is?

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:51]: done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:55]: Done.

File: f3/types.go
Comment (ranchalp) [reply f3/types.go:10]: Ah ok, I thought about this too but I did not do it precisely to distinguish between different things. But agreed that helps remove boilerplate. Done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:70]: Done.

File: sim/sim.go
Comment (ranchalp) [reply sim/sim.go:41]: Done.

**Comment by ranchalp** [comment]
Just addressed the latest comments

File: f3/powertable.go
Comment (Kubuxu) [inline f3/powertable.go:19]: Q: Why do we build the power table sorted power?
Usage-wise, we will almost always query it by ActorID; events SPs appearing and disappearing are also rarer than an order of the power changing (especially in the long tail).

In general, it seems to me that having the power table have a different representation for granite format and different in the wire format (finality certificates) makes sense. Especially since we will need to optimize the wire format for usage in smart contracts.

Comment (Kubuxu) [reply f3/powertable.go:19]: Ahh, ok, I saw it was also part of granite. It needs to communicate which participants signed it.
One option could be communicating that in the form of RLE+ bitfield of ActorIDs, keeping the power table format specifics out of the granite itself.
https://github.com/filecoin-project/go-bitfield/

File: f3/granite.go
Comment (anorth) [reply f3/granite.go:622]: Remember to replace the code here with a call to your new method

File: f3/powertable.go
Comment (anorth) [reply f3/powertable.go:19]: I see what you're saying. Note that the TipSet structure here will have a PowerTable CID added to it, which must match the serialized format used in certificates. However that CID will come from outside, attached to the tipsets pushed in.

The signers will certainly be an RLE+ bitfield (it's present in subsequent PRs). A bitfield over powertable indices will be more dense than over ActorIDs, and I think the density does matter here, as bytes in the messages (of which there are quadratic in the participant count). The sorting by power is also a heuristic to increase density.

These are good questions, lets defer to an issue if you still think it's worth consideration. We can move on here with this, and come back to revise this structure if warranted.

File: f3/powertable.go
Comment (anorth) [reply f3/powertable.go:26]: Sorry, by note I meant "write a comment"

File: go.mod
Comment (anorth) [inline go.mod:16]: It looks like there's some noise crept in here, please normalise before merging.

**Review by anorth — APPROVED** [review]
LGTM after nits

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:19]: We have not gotten there yet (See TODOs in #37's description), but having them sorted by power makes it easy for the creator of a message to optimize message size when creating a justification (i.e. the creator needs to just iterate until the strong quorum is reached when creating the justification). Not visible at the moment even in #37 because all received values that justify the same message are added to the justification (so in many cases this will be more added signatures than required for a strong quorum).

Also the cost of having a sorted power table is negligible compared to running Granite.

File: f3/granite.go
Comment (ranchalp) [reply f3/granite.go:622]: Good catch! Done (upcoming commit)

File: go.mod
Comment (ranchalp) [reply go.mod]: Done.

File: f3/powertable.go
Comment (ranchalp) [reply f3/powertable.go:27]: Done

**Comment by ranchalp** [comment]
Addressed comments from today :)

File: f3/powertable.go
Comment (Kubuxu) [reply f3/powertable.go:19]: Not delaying this PR due to this, as @anorth let's defer this discussion to when we get to finality certificate

**Pull request merged**
